File: frontend/src/api/projects.js
import { apiRequest } from './client.js';

export function listProjects({ page = 1, pageSize = 10 } = {}) {
  return apiRequest(`/projects?page=${page}&page_size=${pageSize}`);
}

export function createProject({ ownerUserId, name, projectType, description, metadata, organizationId, newOrganization, censopasStudy }) {
  return apiRequest('/projects', {
    method: 'POST',
    body: {
      owner_user_id: ownerUserId,
      name,
      project_type: projectType,
      description: description || null,
      metadata: metadata || {},
      organization_id: organizationId ?? null,
      new_organization: newOrganization ?? null,
      censopas_study: censopasStudy ?? null,
    },
  });
}

export function getProject(projectId) {
  return apiRequest(`/projects/${projectId}`);
}

export function updateProject(projectId, payload) {
  return apiRequest(`/projects/${projectId}`, { method: 'PATCH', body: payload });
}

File: backend/app/api/v1/projects.py
from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.pagination import Page, PageParams, page_params
from app.core.security import get_current_user
from app.models.user import OrganizationMembership, User
from app.schemas.projects import ProjectCreate, ProjectRead, ProjectUpdate
from app.services.project_service import ProjectService

router = APIRouter(tags=["projects"])


@router.post("/projects", response_model=ProjectRead, status_code=201)
async def create_project(
    payload: ProjectCreate,
    session: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # El propietario nunca se toma del cuerpo (un cliente no debe poder crear
    # proyectos a nombre de otro usuario) — se deriva siempre del JWT.
    overrides: dict = {"owner_user_id": current_user.id}
    if payload.organization_id is None:
        membership = (
            (
                await session.execute(
                    select(OrganizationMembership).where(
                        OrganizationMembership.user_id == current_user.id
                    )
                )
            )
            .scalars()
            .first()
        )
        if membership is not None:
            overrides["organization_id"] = membership.organization_id

    service = ProjectService(session)
    project = await service.create(payload.model_copy(update=overrides))
    return ProjectRead.model_validate(project)


@router.get("/projects", response_model=Page[ProjectRead])
async def list_projects(
    params: PageParams = Depends(page_params), session: AsyncSession = Depends(get_db)
):
    service = ProjectService(session)
    return await service.list(params)


@router.get("/projects/{project_id}", response_model=ProjectRead)
async def get_project(project_id: int, session: AsyncSession = Depends(get_db)):
    service = ProjectService(session)
    project = await service.get(project_id)
    return ProjectRead.model_validate(project)


@router.patch("/projects/{project_id}", response_model=ProjectRead)
async def update_project(
    project_id: int, payload: ProjectUpdate, session: AsyncSession = Depends(get_db)
):
    service = ProjectService(session)
    project = await service.update(project_id, payload)
    return ProjectRead.model_validate(project)
